### PR TITLE
Java codegen

### PR DESCRIPTION
**Pull request by aslakhellesoy**
# Description

Generate `CiServers.java` at build time, using Groovy as a template engine and the root `ciDict.json` as data for the template. This addresses #17 for the Java implementation.

# Motivation & context

With this change the library no longer needs to parse JSON at runtime. This makes the library:
* more performant (not really important)
* smaller memory footprint (not really important)
* more secure and less maintenance needed, because 3rd party JSON parsers tend to have CVEs

With this change there is no need to copy the `ciDict.json` file either.

## Type of change

- Refactoring/debt (improvement to code design or tooling without changing behaviour)
- Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

File: java/src/main/groovy/io/cucumber/createmeta/CiServers.gsp [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package io.cucumber.createmeta;
+
+import java.util.Arrays;
+
+public class CiServers {
+    public static final Iterable<? extends CiTemplate> TEMPLATES = Arrays.asList(
```
Comment (mpkorstanje) [inline java/src/main/groovy/io/cucumber/createmeta/CiServers.gsp:6]: `Iterable<CiTemplate>` should be fine here.

File: java/src/main/java/io/cucumber/createmeta/CiTemplate.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
+package io.cucumber.createmeta;
+
+public class CiTemplate {
```
Comment (mpkorstanje) [inline java/src/main/java/io/cucumber/createmeta/CiTemplate.java:3]:
```suggestion
public final class CiTemplate {
```

File: java/src/main/java/io/cucumber/createmeta/CodeGenerator.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+import java.io.File;
+import java.io.IOException;
+import java.text.MessageFormat;
+import java.util.Map;
+import java.util.SortedMap;
+import java.util.stream.Collectors;
+
+// https://newbedev.com/generating-sources-by-running-a-project-s-java-class-in-maven
+public class CodeGenerator {
```
Comment (mpkorstanje) [inline java/src/main/java/io/cucumber/createmeta/CodeGenerator.java:16]:
```suggestion
public final class CodeGenerator {
```

File: java/src/main/java/io/cucumber/createmeta/CodeGenerator.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    @JsonIgnoreProperties(ignoreUnknown = true)
+    public static class CiDict {
```
Comment (mpkorstanje) [inline java/src/main/java/io/cucumber/createmeta/CodeGenerator.java:18]:
```suggestion
    public static final class CiDict {
```

File: java/src/main/java/io/cucumber/createmeta/CodeGenerator.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        public String url;
+        public String buildNumber;
+        public Git git;
+    }
+
+    public static class Git {
```
Comment (mpkorstanje) [inline java/src/main/java/io/cucumber/createmeta/CodeGenerator.java:24]:
```suggestion
    public static final class Git {
```

File: java/src/main/java/io/cucumber/createmeta/CiTemplate.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    public final GitTemplate git;
+
+    public CiTemplate(String name, String url, String buildNumber, GitTemplate git) {
+        this.name = name;
+        this.url = url;
+        this.buildNumber = buildNumber;
+        this.git = git;
+    }
+
+    public static class GitTemplate {
```
Comment (mpkorstanje) [inline java/src/main/java/io/cucumber/createmeta/CiTemplate.java:16]:
```suggestion
    public final static class GitTemplate {
```

**Review by mpkorstanje — CHANGES_REQUESTED** [review]
You may want to have a good look at the API you're exposing. Ideally nothing is public unless it is used by a 3rd party (smaller API surface, less to worry about when making changes). Additionally, prefer `final` classes, because it too reduces the API surface.

So I don't think much needs to be public. Made a few suggestions but can't quite check them all,

File: java/src/main/java/io/cucumber/createmeta/CiTemplate.java (hunk 2 above)
Comment (aslakhellesoy) [reply java/src/main/java/io/cucumber/createmeta/CiTemplate.java]: Fixed

File: java/src/main/java/io/cucumber/createmeta/CodeGenerator.java (hunk 3 above)
Comment (aslakhellesoy) [reply java/src/main/java/io/cucumber/createmeta/CodeGenerator.java]: The entire CodeGenerator class was an experiment that I forgot to delete. It's deleted now.

File: java/src/main/java/io/cucumber/createmeta/CiTemplate.java (hunk 6 above)
Comment (aslakhellesoy) [reply java/src/main/java/io/cucumber/createmeta/CiTemplate.java]: Made it `final static`

**Comment by aslakhellesoy** [comment]
Thanks for the feedback @mpkorstanje !

**Pull request merged**